### PR TITLE
Fix asset backfill issue

### PR DESCRIPTION
**Pull request by OwenKephart**
### Summary & Motivation

Resolves: https://github.com/dagster-io/dagster/issues/9287

Basically, when you launched a backfill for assets selected from the AssetGraph, this used to be translated into a set of step_keys_to_execute. There's no precise mapping between step keys and asset keys, so this would fail in more complex situations, such as with multi assets or with reused node definitions.

This change adds in an asset_selection argument to the backfill, essentially allowing you to backfill a set of assets out of a job, instead of a set of steps.

### How I Tested These Changes

The new test failed before these changes (the part where we check the number of run ids for asset keys would fail, because all the runs would materialize non-selected asset keys). Now it passes

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/assets_job.py [hunk 1]
```diff
@@ -231,12 +231,18 @@ def build_node_deps(
 def _has_cycles(deps: Dict[Union[str, NodeInvocation], Dict[str, IDependencyDefinition]]) -> bool:
     """Detect if there are cycles in a dependency dictionary."""
     try:
-        node_deps: Dict[Union[str, NodeInvocation], Set[str]] = {}
+        node_deps: Dict[str, Set[str]] = {}
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/assets_job.py:234]: this change is unrelated -- it fixes a bug that I found when creating the example for the test.

Basically you would get the wrong error type when checking for cycles sometimes because toposort doesn't play nicely with keys that can be multiple different types.

**Comment by sryza** [comment]
I'm going to defer to you on this one @clairelin135

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py [hunk 2]
```diff
@@ -67,6 +68,12 @@ def create_and_launch_partition_backfill(graphene_info, backfill_params):
         reexecution_steps=backfill_params.get("reexecutionSteps"),
         tags={t["key"]: t["value"] for t in backfill_params.get("tags", [])},
         backfill_timestamp=pendulum.now("UTC").timestamp(),
+        asset_selection=[
+            AssetKey.from_graphql_input(asset_key)
+            for asset_key in backfill_params.get("assetSelection")
+        ]
+        if backfill_params.get("assetSelection")
```
Comment (clairelin135) [inline python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:75]: I think it shouldn't be possible for `assetSelection` to have type None if the input param type is
```
graphene.List(graphene.NonNull(GrapheneAssetKeyInput))
```
Would it make sense to make assetSelection take in `None` to signify an asset selection wasn't provided?

**Review by clairelin135 — COMMENTED** [review]
The asset selection stuff looks great to me. Had a couple questions in inline comments

File: python_modules/dagster/dagster/_core/execution/backfill.py [hunk 3]
```diff
@@ -72,6 +75,7 @@ def __new__(
         backfill_timestamp: float,
         last_submitted_partition_name: Optional[str] = None,
         error: Optional[SerializableErrorInfo] = None,
+        asset_selection: Optional[FrozenSet[AssetKey]] = None,
```
[the remark this answers is not in the document]
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/backfill.py:78]: Think this might have to be a list instead?

**Review by clairelin135 — APPROVED** [review]
This is looking great to me! Added one small question, but otherwise 🚢
